agent/submatview/materializer.go

```diff
@@ -0,0 +1,288 @@
+	Subscribe(ctx context.Context, in *pbsubscribe.SubscribeRequest, opts ...grpc.CallOption) (pbsubscribe.StateChangeSubscription_SubscribeClient, error)
+}
+
+// NewMaterializer returns a new Materializer. Run must be called to start it.
+func NewMaterializer(deps Deps) *Materializer {
+	v := &Materializer{
+		deps:        deps,
+		view:        deps.View,
+		retryWaiter: deps.Waiter,
+	}
+	v.reset()
+	return v
+}
+
+// Run receives events from the StreamClient and sends them to the View. It runs
+// until ctx is cancelled, so it is expected to be run in a goroutine.
+func (m *Materializer) Run(ctx context.Context) {
+	for {
+		req := m.deps.Request(m.index)
+		err := m.runSubscription(ctx, req)
+		if ctx.Err() != nil {
+			return
+		}
+
+		failures := m.retryWaiter.Failures()
+		if isNonTemporaryOrConsecutiveFailure(err, failures) {
+			m.lock.Lock()
+			m.notifyUpdateLocked(err)
+			m.lock.Unlock()
+		}
+
+		m.deps.Logger.Error("subscribe call failed",
+			"err", err,
+			"topic", req.Topic,
+			"key", req.Key,
+			"failure_count", failures+1)
+
+		if err := m.retryWaiter.Wait(ctx); err != nil {
+			return
+		}
+	}
+}
+
+// isNonTemporaryOrConsecutiveFailure returns true if the error is a temporary error
+// or if the failures > 0.
+func isNonTemporaryOrConsecutiveFailure(err error, failures int) bool {
+	// temporary is an interface used by net and other std lib packages to
+	// show error types represent temporary/recoverable errors.
+	_, ok := err.(interface {
+		Temporary() bool
+	})
+	return !ok || failures > 0
+}
+
+// runSubscription opens a new subscribe streaming call to the servers and runs
+// for it's lifetime or until the view is closed.
+func (m *Materializer) runSubscription(ctx context.Context, req pbsubscribe.SubscribeRequest) error {
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	m.handler = m.initialHandler(req.Index)
+
+	s, err := m.deps.Client.Subscribe(ctx, &req)
+	if err != nil {
+		return err
+	}
+
+	for {
+		event, err := s.Recv()
+		switch {
+		case isGrpcStatus(err, codes.Aborted):
+			m.reset()
+			return resetErr("stream reset requested")
+		case err != nil:
+			return err
+		}
+
+		m.handler, err = m.handler(event)
+		if err != nil {
+			m.reset()
+			return err
+		}
+	}
+}
+
+func isGrpcStatus(err error, code codes.Code) bool {
+	s, ok := status.FromError(err)
+	return ok && s.Code() == code
+}
+
+// resetErr represents a server request to reset the subscription, it's typed so
+// we can mark it as temporary and so attempt to retry first time without
+// notifying clients.
+type resetErr string
+
+// Temporary Implements the internal Temporary interface
+func (e resetErr) Temporary() bool {
+	return true
+}
+
+// Error implements error
+func (e resetErr) Error() string {
+	return string(e)
+}
+
+// reset clears the state ready to start a new stream from scratch.
+func (m *Materializer) reset() {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+
+	m.view.Reset()
+	m.index = 0
+	m.notifyUpdateLocked(nil)
```

Comment:
Calling `notifyUpdateLocked` here seems strange to me.  Why would we notify on a `reset()` instead of when there are new values? All  the tests fail with this removed, so it must be necessary.

This is a reminder to myself to look into why.

Reply:
I tracked this down. It seems to only be necessary as a way to set `updateCh` to the initial value when `reset()` is called from `NewMaterializer`. Instead I have removed both this call from `reset()` and the call to `reset()` from `New()` and set a default value. This should remove the need to check if `chUpdate` is nil (it never is, because the constructor sets the default value), and should ensure that `Fetch()` is never notified unless there is an error, or an updated value.